### PR TITLE
PMM-5738 Very basic core

### PR DESCRIPTION
**Pull request by percona-csalguero**
- Sandbox
- Connection tests
- Basic metric collecting and renaming

### CODE REVIEW

File: docker/test/init-cluster.sh [hunk 1]
```diff
@@ -0,0 +1,219 @@
+#!/bin/bash
```
Comment (BupycHuk) [inline docker/test/init-cluster.sh:1]: please include a link to the task to replace it with go code

File: exporter/diagnostic_data_collector_test.go [hunk 2]
```diff
@@ -26,3 +32,48 @@ func TestDiagnosticDataCollector(t *testing.T) {
...
+	username := os.Getenv("TEST_MONGODB_ADMIN_USERNAME")
+	password := os.Getenv("TEST_MONGODB_ADMIN_PASSWORD")
+	port := os.Getenv("TEST_MONGODB_S1_PRIMARY_PORT")
+	username = "admin"
+	password = "admin123456"
+	port = "17001"
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	dsn := fmt.Sprintf("mongodb://%s:%s@%s:%s/admin", username, password, hostname, port)
```
Comment (BupycHuk) [inline exporter/diagnostic_data_collector_test.go]: let's extract DNS generating to some utils package

File: exporter/diagnostic_data_collector_test.go [hunk 3]
```diff
@@ -26,3 +32,48 @@ func TestDiagnosticDataCollector(t *testing.T) {
...
+	var obj bson.M
+
+	buf, err = ioutil.ReadFile(path.Join("../internal", "renamer", "testdata", "test001_src.json"))
+	assert.NoError(t, err)
+	err = bson.Unmarshal(buf, &obj)
+	assert.NoError(t, err)
+
+	mm := c.makeMetrics("", obj)
+
+	pretty.Println(mm)
```
Comment (BupycHuk) [inline exporter/diagnostic_data_collector_test.go]: this test doesn't look like a good one

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: docker/test/init-cluster.sh (hunk 1 above)
Comment (BupycHuk) [reply docker/test/init-cluster.sh:1]: https://jira.percona.com/browse/PMM-5753

**Review by AlekSi — CHANGES_REQUESTED** [review]
Please replace all instances of "Percona backup"

**Comment by AlekSi** [comment]
Another thing I think we should add now is CI. Let's try GitHub Actions – we are having a great experience with them on SaaS side. @artemgavrilov can provide an initial configuration and help with them in general.

File: docker/test/init-cluster.sh (hunk 1 above)
Comment (BupycHuk) [reply docker/test/init-cluster.sh:1]: not addressed yet

**Comment by BupycHuk** [comment]
@percona-csalguero Merge conflicts, comment from @AlekSi which is not answered nor addressed. And please don't forgot to re-request review if it should be reviewed.

File: exporter/diagnostic_data_collector.go [hunk 4]
```diff
@@ -21,7 +23,8 @@ func (d *diagnosticDataCollector) Describe(ch chan<- *prometheus.Desc) {
 }
 
 func (d *diagnosticDataCollector) Collect(ch chan<- prometheus.Metric) {
-	cmd := bson.D{{"getDiagnosticData", "1"}}
+	// cmd := bson.D{{Key: "replSetGetStatus", Value: "1"}}
```
Comment (BupycHuk) [inline exporter/diagnostic_data_collector.go]: please remove this comment

File: exporter/exporter_test.go [hunk 5]
```diff
@@ -14,13 +17,61 @@ func getTestClient(t *testing.T) *mongo.Client {
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```
Comment (BupycHuk) [inline exporter/exporter_test.go]: looks like this function shouldn't be in this file.

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+GORELEASER_FLAGS?=
+UID?=$(shell id -u)
+
+export TEST_PSMDB_VERSION?=3.6
+export TEST_MONGODB_FLAVOR?=percona/percona-server-mongodb
+export TEST_MONGODB_ADMIN_USERNAME?=admin
+export TEST_MONGODB_ADMIN_PASSWORD?=admin123456
+export TEST_MONGODB_USERNAME?=test
+export TEST_MONGODB_PASSWORD?=123456
+export TEST_MONGODB_S1_RS?=rs1
```
Comment (BupycHuk) [inline Makefile:27]: any reason to have so much env variables?

File: exporter/metrics.go [hunk 7]
```diff
@@ -42,17 +65,29 @@ func makeRawMetric(name string, value interface{}) (prometheus.Metric, error) {
 		f = float64(v)
 	case primitive.Timestamp:
```
Comment (BupycHuk) [inline exporter/metrics.go]: please use case list from https://yourbasic.org/golang/switch-statement/ for the cases which should be ignored.

File: exporter/metrics_test.go [hunk 8]
```diff
@@ -1,15 +1,9 @@
...
-)
-
-func TestMakeRawMetricName(t *testing.T) {
-	for name, fqName := range map[string]string{
-		"dd.serverStatus.opcountersRepl.delete": "dd_server_status_opcounters_repl_delete",
-	} {
-		assert.Equal(t, fqName, makeRawMetricName(name))
-	}
-}
+//func TestMakeRawMetricName(t *testing.T) {
```
Comment (BupycHuk) [inline exporter/metrics_test.go]: Commented code should not be pushed to production

File: main.go [hunk 9]
```diff
@@ -1,11 +1,58 @@
...
+	if opts.Debug {
+		logrus.SetLevel(logrus.DebugLevel)
+	}
+
+	exporterOpts := &exporter.Opts{
+		DSN: opts.DSN,
+		Log: log,
+	}
+
+	e, err := exporter.New(exporterOpts)
```
Comment (BupycHuk) [inline main.go]: @percona-csalguero Do we have task for API? looks like we missed it.

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: Makefile (hunk 6 above)
Comment (percona-csalguero) [reply Makefile:27]: For the sandbox, we need 3 servers for RS1, 3 for RS2, 1 config server, 1 mongos server and one stand alone server. via this variables we can set user, pass, rs names and ports for each one.

File: main.go (hunk 9 above)
Comment (percona-csalguero) [reply main.go]: No. We have independent tasks for each collector

File: .github/workflows/go.yml [hunk 10]
```diff
@@ -24,13 +24,19 @@ jobs:
 
     - name: Get dependencies
       run: |
+        make init
         go get -v -t -d ./...
         if [ -f Gopkg.toml ]; then
```
Comment (BupycHuk) [inline .github/workflows/go.yml]: do we need this check?

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    - docker-compose --version
+
+install:
+  # install reviewdog and golangci-lin
+  - curl -sfL https://raw.githubusercontent.com/reviewdog/reviewdog/master/install.sh| sh -s
+  - curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s latest
+
+before_script:
+  # static analyze
+  - bin/golangci-lint run -c=.golangci-required.yml --out-format=line-number | bin/reviewdog -f=golangci-lint -level=error -reporter=github-pr-check
```
Comment (BupycHuk) [inline .travis.yml]: .golangci-required.yml is not presented in this PR

File: .travis.yml [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+  - bin/golangci-lint run -c=.golangci.yml --out-format=line-number | bin/reviewdog -f=golangci-lint -level=error -reporter=github-pr-review
```
Comment (BupycHuk) [inline .travis.yml]: To make it work we need REVIEWDOG_GITHUB_API_TOKEN environment variable

File: .travis.yml [hunk 13]
```diff
@@ -0,0 +1,74 @@
+language: go
```
Comment (BupycHuk) [inline .travis.yml]: Do we really need github actions and travis ci in the same time?

File: exporter/exporter_test.go [hunk 14]
```diff
@@ -2,25 +2,80 @@ package exporter
...
-	t.Cleanup(func() { client.Disconnect(ctx) })
+	t.Cleanup(func() { client.Disconnect(ctx) }) //nolint:errcheck
 
 	err = client.Ping(ctx, nil)
 	require.NoError(t, err)
 
 	return client
 }
+
+func TestConnect(t *testing.T) {
```
Comment (BupycHuk) [inline exporter/exporter_test.go]: I think we should ignore this kind of things in linter.

File: main.go (hunk 9 above)
Comment (BupycHuk) [reply main.go]: But, do we have task for API core? Because I don't see any ability to get these metrics by prometheus.

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: main.go (hunk 9 above)
Comment (percona-csalguero) [reply main.go]: In https://github.com/Percona-Lab/mnogo_exporter/pull/3

    Diagnostic data collector
    Replicaset status collector
    Collstats collector
    Server status collector
    Main parameters
    HTTP server that will support having multiple paths to expose metrics

File: .travis.yml (hunk 13 above)
Comment (percona-csalguero) [reply .travis.yml]: deleted

File: main.go (hunk 9 above)
Comment (BupycHuk) [reply main.go]: please include the link to task about API here

File: .travis.yml (hunk 13 above)
Comment (BupycHuk) [reply .travis.yml]: it's weird, but it's not deleted :)

File: main.go [hunk 15]
```diff
@@ -1,13 +1,63 @@
 package main
 
 import (
+	"fmt"
+
 	"github.com/Percona-Lab/mnogo_exporter/exporter"
```
Comment (BupycHuk) [inline main.go]: Please, fix goimports like we do in other repos.

**Comment by BupycHuk** [comment]
Can we remove all unrelated code?

**Comment by BupycHuk** [comment]
Tests are failed

**Comment by percona-csalguero** [comment]
Closing in favor of https://github.com/Percona-Lab/mnogo_exporter/pull/4

**Pull request closed without merging**